### ISSUE TITLE
Clippy fails with "toolchain does not have the binary `cargo-clippy`"

### ISSUE THREAD

**Issue by matthias-t** [CAD97/great-ci#8]
After I ran into a problem with a similar config, I forked this repository and triggered [a build](https://travis-ci.com/matthias-t/great-ci/jobs/135381890). In the clippy and rustfmt job (`.4`), the command `cargo clippy -- -D clippy` exits with an error:
```go
error: toolchain 'nightly-2018-06-18-x86_64-unknown-linux-gnu' does not have the binary `cargo-clippy`
```

**Comment by matthias-t** [comment]
Taking a look at the rust-clippy `README.md`'s history (see rust-lang-nursery/rust-clippy#2921), it looks like the install method changed to:
```
rustup component add clippy-preview
```

**Comment by CAD97** [comment]
👍 See also https://internals.rust-lang.org/t/clippy-is-available-as-a-rustup-component/7967/16?u=cad97

> What seems to be happening is that rustup has provided a cargo-clippy binary already, so the cache-friendly install of clippy sees that binary there and doesn’t install from crates.io. Then trying to use it obviously fails because the binary just says that the current toolchain doesn’t have clippy.

I'll leave this issue open as an indicator of the change. Since clippy's riding the trains to stable, I want to update the blog post once it's there and we've got Edition 2018 support on stable. Basically an evolution the current blog post -- how to set up a new project with as much of the somewhat standard tooling as possible.

Finally, see also https://crate-ci.github.io/ where me and \@epage are working to provide a somewhat consistent book of good practices.

**Comment by epage** [comment]
@CAD97 I'm curious, is there anything left here that hasn't been merged into crate-ci's book or examples?

**Comment by matthias-t** [comment]
@CAD97 thanks for the links! I'm happy that you are working on this book. Managing CI & co. can be a pain, especially when you just want to code, not fight with configuration. I managed to fix my CI thanks to crate-ci/example-clippy@fa821f3486ff18bc5feaf46ae2d0e1a23e744c45.

Also, I've looked through the rustup docs (their `README.md`, apparently), but couldn't find any information on toolchain components. 
- What does it mean for clippy to be a rustup component (compared to a regular crate)? 
- How does it prevent nightly breakages?

Keep me posted on that blog post!

**Comment by epage** [comment]
Clippy now ships with rust
- faster ci because it comes prebuilt
- no more guessing what version of clippy matches with what nightly version
- "fixes" CIs for people who don't pin nightly+clippy since now every nightly will have a clippy. Before, if the nightly broke clippy, there was a window before clippy was fixed
  - for now you should still pin them because new clippy warnings could break your build (contributors PRs breaking for problems unrelated to them is discouraging)

**Comment by matthias-t** [comment]
@epage that's enlightening, thanks a lot!

**Comment by matthias-t** [comment]
I think it landed in stable rust

**Comment by CAD97** [comment]
Yep! I've got a couple small drafts put together; I'll probably have something more concrete as 1.31 reaches beta. I think the successor will be "Setting up a Rust project for Success" and it should cover basically everything around both CI and tooling for testing as it stands at the head of Edition 2018.


I'll try and share a draft here once I've started actually writing it. For now, `rustup component add clippy-preview` and never worry about building it yourself again :)
